File: apps/desktop/packages/server-api-contract/src/envelope.ts
import { create, type MessageInitShape } from "@bufbuild/protobuf";
import { type Envelope, EnvelopeSchema } from "./gen/kent/api/shared/foundation_pb.js";
import { decode, decodeUnvalidated, encode } from "./message.js";

export function decodeEnvelope(bytes: Uint8Array): Envelope {
  return decode(EnvelopeSchema, bytes);
}

export function decodeEnvelopeCorrelation(bytes: Uint8Array): string | undefined {
  let envelope: Envelope;
  try {
    envelope = decodeUnvalidated(EnvelopeSchema, bytes);
  } catch {
    return undefined;
  }
  const correlation =
    envelope.frame.case === "call" ||
    envelope.frame.case === "result" ||
    envelope.frame.case === "transportFailure"
      ? envelope.frame.value.correlation
      : undefined;
  return correlation === undefined || correlation.trim().length === 0 ? undefined : correlation;
}

export function encodeEnvelope(envelope: MessageInitShape<typeof EnvelopeSchema>): Uint8Array {
  return encode(EnvelopeSchema, create(EnvelopeSchema, envelope));
}

File: shared/protoapi/envelope.go
package protoapi

import (
	"fmt"
	"strings"

	sharedpb "core/shared/protoapi/gen/kent/api/shared"
)

func DecodeEnvelope(encoded []byte) (*sharedpb.Envelope, error) {
	envelope := &sharedpb.Envelope{}
	if err := Decode(encoded, envelope); err != nil {
		return nil, fmt.Errorf("decode envelope: %w", err)
	}
	return envelope, nil
}

// DecodeEnvelopeCorrelation recovers only a syntactically decodable correlation
// so callers can isolate an invalid frame without accepting the envelope.
func DecodeEnvelopeCorrelation(encoded []byte) string {
	envelope := &sharedpb.Envelope{}
	if err := unmarshalGenerated(encoded, envelope); err != nil {
		return ""
	}
	var correlation *string
	switch frame := envelope.GetFrame().(type) {
	case *sharedpb.Envelope_Call:
		correlation = frame.Call.Correlation
	case *sharedpb.Envelope_Result:
		correlation = frame.Result.Correlation
	case *sharedpb.Envelope_TransportFailure:
		correlation = frame.TransportFailure.Correlation
	}
	if correlation == nil || strings.TrimSpace(*correlation) == "" {
		return ""
	}
	return *correlation
}

func EncodeEnvelope(envelope *sharedpb.Envelope) ([]byte, error) {
	encoded, err := Encode(envelope)
	if err != nil {
		return nil, fmt.Errorf("encode envelope: %w", err)
	}
	return encoded, nil
}
